webapp/src/components/common/MatrixGrid/useMatrix.ts

```diff
@@ -53,27 +72,72 @@ export function useMatrix(
   const [isSubmitting, setIsSubmitting] = useState(false);
   const [error, setError] = useState<Error | undefined>(undefined);
   const [{ present: currentState }, { set: setState, undo, redo, canRedo }] =
-    useUndo<DataState>({ data: [], pendingUpdates: [] });
+    useUndo<DataState>({
+      data: [],
+      aggregates: { min: [], max: [], avg: [], total: [] },
+      pendingUpdates: [],
+      updateCount: 0,
+    });
 
-  const fetchMatrix = useCallback(async () => {
-    setIsLoading(true);
-    try {
-      const [matrix, index] = await Promise.all([
-        getStudyData<MatrixDataDTO>(studyId, url),
-        getStudyMatrixIndex(studyId, url),
-      ]);
-
-      setState({ data: matrix.data, pendingUpdates: [] });
-      setColumnCount(matrix.columns.length);
-      setIndex(index);
-      setIsLoading(false);
-    } catch (error) {
-      setError(new Error(t("data.error.matrix")));
-      enqueueErrorSnackbar(t("data.error.matrix"), error as AxiosError);
-    } finally {
-      setIsLoading(false);
-    }
-  }, [enqueueErrorSnackbar, setState, studyId, url]);
+  // Determine the aggregate types to display in the matrix
+  const aggregateTypes = useMemo(
+    () => getAggregateTypes(aggregatesConfig || []),
+    [aggregatesConfig],
+  );
+
+  // Display warning prompts to prevent unintended navigation
+  // 1. When the matrix is currently being submitted
+  usePrompt(t("form.submit.inProgress"), isSubmitting);
+  // 2. When there are unsaved changes in the matrix
+  usePrompt(t("form.changeNotSaved"), currentState.pendingUpdates.length > 0);
+
+  const fetchMatrix = useCallback(
+    async (loadingState = true) => {
+      // !NOTE This is a temporary solution to ensure the matrix is up to date
+      // TODO: Remove this once the matrix API is updated to return the correct data
+      if (loadingState) {
+        setIsLoading(true);
+      }
+
+      try {
+        const [matrix, index] = await Promise.all([
+          fetchMatrixData
+            ? // If a custom fetch function is provided, use it
+              fetchMatrixData(studyId)
+            : getStudyData<MatrixDataDTO>(studyId, url, 1),
+          getStudyMatrixIndex(studyId, url),
+        ]);
+
+        setState({
+          data: matrix.data,
+          aggregates: calculateMatrixAggregates(matrix.data, aggregateTypes),
+          pendingUpdates: [],
+          updateCount: 0,
+        });
+        setColumnCount(matrix.columns.length);
+        setIndex(index);
+        setIsLoading(false);
+
+        return {
+          matrix,
+          index,
+        };
+      } catch (error) {
+        setError(new Error(t("data.error.matrix")));
+        enqueueErrorSnackbar(t("data.error.matrix"), error as AxiosError);
+      } finally {
+        setIsLoading(false);
+      }
+    },
+    [
+      aggregateTypes,
+      enqueueErrorSnackbar,
+      fetchMatrixData,
+      setState,
+      studyId,
+      url,
+    ],
+  );
 
   useEffect(() => {
     fetchMatrix();
```

Comment:
Better to use deps here instead of creating a `useCallback` for `fetchMatrix`

Reply:
yes but its used elsewhere in the hook that's why

Reply:
Looking at it closer `useCallback` is indeed not necessary. A brief explanation like "not needed since function is only invoked internally, not used as prop/dependency" would make the review more valuable.